README.md
# fluent-plugin-everysense
Fluent Input/Output Plugin for EverySense platform

## What is EverySense?

Collecting personal data is one of the key points in IoT applications. However, it is basically difficult to search people who can provide own sensor data and to motivate people for providing own data to the world. [EverySense](http://every-sense.com/) is a platform to exchange sensor data which can define own recipe to search the data provider matching the condition. It also provide "point" mechanism to motivate people for providing own data. The "point" can be exchanged with the other sensor data or goods like general creadit card point.

This plugin provides input/output function to [EverySense platform](https://service.every-sense.com/). Data from fluentd event can be uploaded to EverySense and the data provided from EverySense can be posted into fluentd network.


## Installation

```
gem install fluent-plugin-everysense
```


## Usage

fluent-plugin-everysense has Input and Output Plugins for EverySense platform.


### Input Plugin (Fluent::EverySenseInput)

Input Plugin can receive events from EverySense Server. It can be used via source directive in the configuration.

```
<source>
  @type everysense
  format json
  tag tag_name
  login_name your_login_name
  password your_password
  recipe_id xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx
  polling_interval 30
</source>
```

- **format** (required): type of parser can be specified to parse input data. currently only json format is supported. parser plugin options can also be specified.
- **tag** (required): tag name appended to the input data inside fluentd network
- **login_name** (required): login_name to login https://service.every-sense.com/
- **password** (required): password for the login_name
- **device_id** (device_id or recipe_id is required): the target device id to obtain input data
- **recipe_id** (device_id or recipe_id is required): the target recipe id to obtain input data
- **polling_interval**: interval to poll EverySense JSON over HTTP API

Since each device may have multiple sensors, time field is generated when Input Plugin received data from EverySense. It can be used only for inside the fluentd network because the real timestamps for the sensors are recorded inside JSON data and not synchronized to the time field. An example record format of the Input Plugin is as follows:

```
{"farm_uuid":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx",
 "device":
  [
    { "data": {
        "at": "2016-05-15 12:14:30 +0900",
        "unit":"degree Celsius",
        "value":23
      },
      "sensor_name":"collection_data_1"
    },
    { "data": {
        "at":"2016-05-15 12:14:30 +0900",
        "unit":"%RH",
        "value":30
      },
      "sensor_name":"collection_data_2"
    }
  ]
}
```

While fluentd record must be a map (Hash), the output of EverySense becomes an array. So, a key ("json") is added to make it as a map.

### Output Plugin (Fluent::EverySenseOutput)

Output Plugin can send events to EverySense server. It can be used via match directive. Output Plugin assumes the input format as described above.

```
<match tag_name>
  @type everysense
  login_name your_login_name
  password your_password
  device_id xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx
  flush_interval 10
  aggr_type none
  <sensor>
    input_name collection_data_1
    output_name FESTIVAL_Test1_Sensor
    type_of_value Integer
  </sensor>
  <sensor>
    input_name collection_data_2
    output_name FESTIVAL_Test1_Sensor2
    type_of_value Integer
  </sensor>
</match>
```

- **login_name** (required): login_name to login https://service.every-sense.com/
- **password** (required): password for the login_name
- **device_id** (required): The target device id to submit sensor data
- **flush_interval**: Upload interval (default: 30)
- **aggr_type**: Buffered sensor data can be aggregated with specified statistic method. Currently avg (average) and none (without aggregation) can be specified (default: none).
- **sensor** (required): Output sensor names must be specified by sensor directive. Sensor data with the specified input_name will be uploaded to EverySense server with the output_name. If the input device data includes multiple sensor data, multiple directives can be specified. If sensor_name of input sensor data is not specified in the configuration, that sensor data will be skipped. If the sensor_names of input data are nil, sensor data will be uploaded with the number of sensor directives with the specified output_name.
  - **input_name**: sensor_name of the input sensor data.
  - **output_name**: sensor_name of the output sensor data.
  - **type_of_value**: type of value (default: Integer)

## Filter Plugin (Fluent::EverySenseFilter)

Filter Plugin can split an EverySense Server event into multiple fluentd events if it has multiple data. It can be used via filter directive.

```
<filter tag_name>
  @type everysense
</filter>
```

The following input data from EverySense Server will be splitted into multiple fluentd events.

```
{"json":
  [
    { "farm_uuid":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx",
      "sensor_name":"collection_data_1",
      "data_class_name":"Illuminance",
      "data": {
        "at":"2016-08-24 00:15:00 UTC",
        "value":137.0,
        "unit":"lx"
      }
    },
    { "farm_uuid":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx",
      "sensor_name":"collection_data_2",
      "data_class_name":"Location",
      "data": {
        "at":"2016-08-24 00:15:00 UTC",
        "values":[138.442062,35.8162422,671.599975],
        "unit":"degree"
      }
    }
  ]
}
```

```
{ "farm_uuid":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx",
  "sensor_name":"collection_data_1",
  "data_class_name":"Illuminance",
  "data": {
    "at":"2016-08-24 00:15:00 UTC",
    "value":137.0,
    "unit":"lx"
  }
},
{ "farm_uuid":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx",
  "sensor_name":"collection_data_2",
  "data_class_name":"Location",
  "data": {
    "at":"2016-08-24 00:15:00 UTC",
    "values":[138.442062,35.8162422,671.599975],
    "unit":"degree"
  }
}
```

## Contributing

1. Fork it ( http://github.com/toyokazu/fluent-plugin-everysense/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


## License

The gem is available as open source under the terms of the [Apache License Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).


CHANGELOG.md
## 0.0.8, 0.1.0 (2017-04-01)

Features:
  - compatible with fluentd v0.14
  - change license from MIT License to Apache License Version 2.0

Deprecations:

  - deprecated support of fluentd v0.12


lib/fluent/plugin/in_everysense.rb
require 'fluent/plugin/input'
require 'fluent/event'
require 'fluent/time'
require 'fluent/plugin/everysense_proxy'

module Fluent::Plugin
  class EverySenseInput < Input
    include EverySenseProxy

    Fluent::Plugin.register_input('everysense', self)

    helpers :thread, :compat_parameters, :parser

    desc 'The tag of the event.'
    config_param :tag, :string
    desc 'Polling interval to get message from EverySense API'
    config_param :polling_interval, :integer, :default => 60
    desc 'Device ID'
    config_param :device_id, :string, :default => nil
    desc 'Recipe ID'
    config_param :recipe_id, :string, :default => nil

    # received message format of EverySense is as follows
    #
    # [
    #   [
    #     {
    #       "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #       "sensor_name": "collection_data_1",
    #       "data_class_name": "AirTemperature",
    #       "data": {
    #         "at": "2016-05-12 21:38:52 UTC",
    #         "memo": null,
    #         "value": 23,
    #         "unit": "degree Celsius"
    #       }
    #     },
    #     {
    #       "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #       "sensor_name": "collection_data_2",
    #       "data_class_name": "AirHygrometer",
    #       "data": {
    #         "at": "2016-05-12 21:38:52 UTC",
    #         "memo": null,
    #         "value": 30,
    #         "unit": "%RH"
    #       }
    #     }
    #   ]
    # ]

    def configure(conf)
      super
      configure_parser(conf)
    end

    def configure_parser(conf)
      compat_parameters_convert(conf, :parser)
      parser_config = conf.elements('parse').first
      @parser = parser_create(conf: parser_config)
    end

    def start
      #raise StandardError.new if @tag.nil?
      super
      start_proxy
      @proxy_thread = Thread.new(&method(:get_loop))
    end

    def get_loop
      while (true)
        begin
          messages = get_messages
          emit(messages) if !(messages.nil? || messages.empty?)
          sleep @polling_interval
        rescue Exception => e
          $log.error :error => e.to_s
          $log.debug(e.backtrace.join("\n"))
          #$log.debug_backtrace(e.backtrace)
          sleep @polling_interval
        end
      end
    end

    def parse_time(record)
      if record["data"]["at"].nil?
        $log.debug "Since time_key field is nil, Fluent::EventTime.now is used."
        Fluent::EventTime.now
      else
        @parser.parse_time(record["data"]["at"])
      end
    end

    # Converted record for emission
    # [
    #   {"farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #    "device":
    #     [
    #       {
    #         "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #         "sensor_name": "collection_data_1",
    #         "data_class_name": "AirTemperature",
    #         "data": {
    #           "at": "2016-05-12 21:38:52 UTC",
    #           "memo": null,
    #           "value": 23,
    #           "unit": "degree Celsius"
    #         }
    #       },
    #       {
    #         "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #         "sensor_name": "collection_data_2",
    #         "data_class_name": "AirHygrometer",
    #         "data": {
    #           "at": "2016-05-12 21:38:52 UTC",
    #           "memo": null,
    #           "value": 30,
    #           "unit": "%RH"
    #         }
    #       }
    #     ]
    #   }
    # ]
    def emit(messages)
      begin
        time, record = @parser.parse(messages) do |time, record|
          [time, record]
        end
        #puts "time: #{time.inspect}"
        #puts "record: #{record.inspect}"
        if record.is_a?(Array) && record[0].is_a?(Array) # Multiple devices case
          mes = Fluent::MultiEventStream.new
          record.each do |single_record|
            # use timestamp of the first sensor (single_record[0])
            mes.add(parse_time(single_record[0]), {
              "farm_uuid": single_record[0]["farm_uuid"],
              "device": single_record
            })
          end
          router.emit_stream(@tag, mes)
        elsif record.is_a?(Array) && record[0].is_a?(Hash) # Single device case
          # use timestamp of the first sensor (single_record[0])
          router.emit(@tag, parse_time(record[0]), {
            "farm_uuid": record[0]["farm_uuid"],
            "device": record
          })
        else # The other case
          raise Fluent::Plugin::Parser::ParserError, "Unexpected input format."
        end
      rescue Exception => e
        $log.error :error => e.to_s
        $log.debug_backtrace(e.backtrace)
      end
    end

    def shutdown
      @proxy_thread.kill
      shutdown_proxy
      super
    end
  end
end


lib/fluent/plugin/everysense_proxy.rb
module Fluent::Plugin
  module EverySenseProxy
    require 'uri'
    require 'net/http'
    require 'json'
    require 'time'

    class EverySenseProxyError
    end

    def self.included(base)
      base.desc 'EverySense API URI'
      base.config_param :url, :string, :default => 'https://api.every-sense.com:8001/'
      base.desc 'login_name for EverySense API'
      base.config_param :login_name, :string
      base.desc 'password for EverySense API'
      base.config_param :password, :string
      base.desc 'Maximum number of entries returned'
      base.config_param :limit, :integer, :default => 1000
      #base.config_param :keep_alive, :integer, :default => 2
      base.desc 'Start time of the entries returned'
      base.config_param :from, :string, :default => Time.now.iso8601
      base.config_param :keep, :bool, :default => false
      base.config_param :inline, :bool, :default => false
      base.config_param :format, :string, :default => "json"
    end

    def start_proxy
      $log.debug "start everysense proxy #{@url}"

      @uri = URI.parse(@url)
      @https = Net::HTTP.new(@uri.host, @uri.port)
      @https.use_ssl = (@uri.scheme == 'https')
      @session_key = nil
    end

    def shutdown_proxy
      $log.debug "shutdown_proxy #{@session_key}"
      delete_session
      @https.finish() if @https.active?
    end

    def error_handler(response, message)
      if response.code != "200"
        $log.error :error => message
        $log.debug "code: #{response.code}"
        $log.debug "message: #{response.message}"
        $log.debug "body: #{response.body}"
        return false
      end
      return true
    end

    def valid_session?
      !@session_key.nil? # TODO validate @session_key using EverySense API
    end

    def create_session_request
      session_req = Net::HTTP::Post.new(@uri + '/session')
      session_req.body = {login_name: @login_name, password: @password}.to_json
      session_req.content_type = 'application/json'
      session_req
    end

    def create_session
      return @session_key if valid_session?
      @session_req ||= create_session_request
      session_res = @https.request(@session_req)
      return nil if !error_handler(session_res, 'create_session failed.')
      @session_key = JSON.parse(session_res.body)["session_key"]
    end

    def delete_session_request
      Net::HTTP::Delete.new(@uri + "/session/#{@session_key}")
    end

    def delete_session
      return if !valid_session?
      del_session_res = @https.request(delete_session_request)
      error_handler(del_session_res, 'delete_session failed.')
    end

    def put_message_request(message)
      put_message_req = Net::HTTP::Post.new(@uri + "/device_data/#{@device_id}")
      put_message_req.body = message
      put_message_req.content_type = "application/#{@format}"
      put_message_req
    end

    def put_message(message)
      $log.debug "put_message: #{message}"
      put_message_res = @https.request(put_message_request(message))
      error_handler(put_message_res, "put_message: '#{message}' failed.")
    end

    def target_path
      if !@device_id.nil?
        return "/device_data/#{@device_id}"
      elsif !@recipe_id.nil?
        return "/recipe_data/#{@recipe_id}.#{@format}"
      else
        raise Fluent::ConfigError, "device_id or recipe_id must be specified."
      end
    end

    def get_messages_params
      params = {
        session_key: @session_key,
        from: @from,
        to: Time.now.iso8601,
        limit: @limit
      }
      if !@device_id.nil?
        return params
      elsif !@recipe_id.nil?
        return params.merge({keep: @keep, inline: @inline, format: @format})
      else
        raise ConfigError, "device_id or recipe_id must be specified."
      end
    end

    def get_messages_request
      get_messages_req = @uri + target_path
      get_messages_req.query = URI.encode_www_form(get_messages_params)
      $log.debug "#{get_messages_req}?#{get_messages_req.query}"
      # currently time window is automatically updated
      @from = Time.now.iso8601
      get_messages_req
    end

    def get_messages
      if !valid_session?
        return nil if create_session.nil?
        $log.debug "session #{@session_key} created."
      end
      get_messages_res = @https.get(get_messages_request)
      return nil if !error_handler(get_messages_res,"get_messages failed.")
      $log.debug "get_message: #{get_messages_res.body}"
      get_messages_res.body
    end
  end
end


lib/fluent/plugin/out_everysense.rb
require 'fluent/plugin/output'
require 'fluent/event'
require 'fluent/time'
require 'fluent/plugin/everysense_proxy'

module Fluent::Plugin
  # EverySenseOutput
  # output data to EverySense server
  # this module assumes the input format follows everysense output specification
  class EverySenseOutput < Output
    include EverySenseProxy

    Fluent::Plugin.register_output('everysense', self)

    helpers :compat_parameters, :formatter, :inject

    # config_param defines a parameter. You can refer a parameter via @path instance variable
    # Without :default, a parameter is required.
    desc 'Device ID'
    config_param :device_id, :string
    desc 'Flush interval to put message to EverySense API'
    config_param :flush_interval, :integer, :default => 30
    # <sensor> is mandatory option
    # an example configuraton is shown below
    #
    # <sensor>
    #   input_name collection_data_1
    #   output_name FESTIVAL_Test1_Sensor
    #   type_of_value Integer
    # </sensor>
    # <sensor>
    #   input_name collection_data_2
    #   output_name FESTIVAL_Test1_Sensor2
    #   type_of_value Integer
    # </sensor>
    config_section :sensor, param_name: :sensors, required: true, multi: true do
      desc 'Input sensor name'
      config_param :input_name, :string, :default => nil
      desc 'Output sensor name'
      config_param :output_name, :string
      # type_of_value: "Integer", "Float", "String", etc. The same as Ruby Class name.
      desc 'Type of value'
      config_param :type_of_value, :string, :default => "Integer"
      # unit: "degree Celsius", "%RH", ...
      #desc 'unit'
      #config_param :unit, :string, :default => nil
    end

    # This method is called before starting.
    # 'conf' is a Hash that includes configuration parameters.
    # If the configuration is invalid, raise Fluent::ConfigError.
    def configure(conf)
      super
      compat_parameters_convert(conf, :formatter, :inject, :buffer, default_chunk_key: "time")
      formatter_config = conf.elements(name: 'format').first
      @formatter = formatter_create(conf: formatter_config)
      @has_buffer_section = conf.elements(name: 'buffer').size > 0
      #@sensor.each do |s|
      #  $log.debug s.to_h.inspect
      #end
      @out_sensors = {}
      @sensors.each do |sensor|
        if sensor.input_name.nil?
          @out_sensors[sensor.output_name] = sensor
        else
          @out_sensors[sensor.input_name] = sensor
        end
      end
      $log.debug @out_sensors.inspect
    end

    # This method is called when starting.
    # Open sockets or files here.
    def start
      super
      start_proxy
    end

    def prefer_buffered_processing
      @has_buffer_section
    end

    def force_type(value, out_sensor)
      case out_sensor.type_of_value
      when "Integer"
        return value.to_i
      when "Float"
        return value.to_f
      when "String"
        return value.to_s
      else
        return value.to_i
      end
    end

    # output message format of EverySense is as follows
    #
    # [
    #   {
    #     "data": {
    #       "at":"2016-04-14 17:15:00 +0900",
    #       "unit":"degree Celsius",
    #       "value":23
    #     },
    #     "sensor_name":"FESTIVAL_Test1_Sensor"
    #   },
    #   {
    #     "data": {
    #       "at":"2016-04-14 17:15:00 +0900",
    #       "unit":"%RH",
    #       "value":30
    #     },
    #     "sensor_name":"FESTIVAL_Test1_Sensor2"
    #   }
    # ]
    def transform_in_sensor(in_sensor, out_sensor) # modify sensor_name
      {
        data: {
          at: Time.parse(in_sensor["data"]["at"]),
          unit: in_sensor["data"]["unit"],
          value: force_type(in_sensor["data"]["value"], out_sensor)
        },
        sensor_name: out_sensor.output_name
      }
    end

    def get_out_sensor_by_name(input_name)
      @out_sensors[input_name]
    end

    def get_out_sensor_by_index(index)
      @out_sensors[@out_sensors.keys[index]]
    end

    # Assumed input message format is as follows
    #
    # [
    #   [
    #     {
    #       "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #       "sensor_name": "collection_data_1",
    #       "data_class_name": "AirTemperature",
    #       "data": {
    #         "at": "2016-05-12 21:38:52 UTC",
    #         "memo": null,
    #         "value": 23,
    #         "unit": "degree Celsius"
    #       }
    #     },
    #     {
    #       "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #       "sensor_name": "collection_data_2",
    #       "data_class_name": "AirHygrometer",
    #       "data": {
    #         "at": "2016-05-12 21:38:52 UTC",
    #         "memo": null,
    #         "value": 30,
    #         "unit": "%RH"
    #       }
    #     }
    #   ]
    # ]
    def transform_in_device(in_device)
      if in_device[0]["sensor_name"].nil?
        # if first input data does not include sensor_name,
        # output_names are used in the specified order.
        return in_device.map.with_index do |in_sensor, i|
          if !get_out_sensor_by_index(i).nil?
            transform_in_sensor(in_sensor, get_out_sensor_by_index(i))
          end
        end.compact
      else
        return in_device.map do |in_sensor|
          #$log.debug in_sensor["sensor_name"]
          if @out_sensors.keys.include?(in_sensor["sensor_name"])
            transform_in_sensor(in_sensor, get_out_sensor_by_name(in_sensor["sensor_name"]))
          end
        end.compact
      end
    end

    # Emitted record inside fluentd network
    # [
    #   {"farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #    "device":
    #     [
    #       {
    #         "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #         "sensor_name": "collection_data_1",
    #         "data_class_name": "AirTemperature",
    #         "data": {
    #           "at": "2016-05-12 21:38:52 UTC",
    #           "memo": null,
    #           "value": 23,
    #           "unit": "degree Celsius"
    #         }
    #       },
    #       {
    #         "farm_uuid": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
    #         "sensor_name": "collection_data_2",
    #         "data_class_name": "AirHygrometer",
    #         "data": {
    #           "at": "2016-05-12 21:38:52 UTC",
    #           "memo": null,
    #           "value": 30,
    #           "unit": "%RH"
    #         }
    #       }
    #     ]
    #   }
    # ]
    def put_event_stream(tag, es)
      if es.class == Fluent::OneEventStream
        es = inject_values_to_event_stream(tag, es)
        es.each do |time, record|
          $log.debug "#{tag}, #{record}"
          put_message(@formatter.format(tag, time, transform_in_device(record["device"])))
        end
      else
        es = inject_values_to_event_stream(tag, es)
        array = []
        es.each do |time, record|
          $log.debug "#{tag}, #{record}"
          device = transform_in_device(record["device"])
          array << device if !device.empty?
        end
        put_message(@formatter.format(tag, Fluent::EventTime.now, array))
      end
      $log.flush
    end

    def process(tag, es)
      put_event_stream(tag, es)
    end

    def write(chunk)
      return if chunk.empty?
      tag = chunk.metadata.tag

      put_event_stream(tag, es)
    end

    # This method is called when shutting down.
    # Shutdown the thread and close sockets or files here.
    def shutdown
      shutdown_proxy
      super
    end
  end
end
